sklearn/metrics/_pairwise_distances_reduction/meson.build

```diff
@@ -0,0 +1,161 @@
+_pairwise_distances_reduction_cython_tree = [
+  fs.copyfile('__init__.py'),
+]
```

Comment:
So this things are needed for Cython cimports to work, you generally need `__init__.py` and `.pxd` copied to the build directory, so that Cython detects this as a package and can cimport the relevant files. `_pairwise_distances_reduction` is cimported e.g. in metrics.

This is kind of the reason I named these kind of variables `*_cython_tree` (following numpy and scipy I think).

Honestly I don't know where to put this kind of dark knowledge I gained while putting this together and will probably forget soon. I think numpy and scipy have some comments about this I'll see if I can sprinkle them in the source code or maybe in the developer doc somewhere, not sure.

Reply:
Maybe having a comment directly here would be a good start. At least one reader (Adrin) looked at it and thought "yeah I don't understand this, it looks different from other stuff I've seen". Maybe there are also other places, but this seems like a good place to start.

Reply:
I plan to write this up as Meson docs at some point, since it has explicit Cython support. 

It really belongs in the Cython docs in my opinion, because it'll work the same with CMake or any other build tool that builds out of place. But Cython devs are quite resistant to anything that's build related or not `distutils`/`setuptools`, so I don't have high hopes for that location.

Reply:
Great to hear! In the meantime, I did my best to add some comments about these `*cython_tree*` variables in a recent commit (yesterday or the day before) https://github.com/scikit-learn/scikit-learn/pull/28040/commits/182f1c579a93d363acfce4876231edec82662f3a

Reply:
Yes, those comments all look correct to me. The root cause is that Cython doesn't simply need include/library paths like C/C++, but it walks up the directory tree that the `.pyx` is in, and it derives whether it's part of a Python package or not from `__init__.py` files - and changes the extension module name if it doesn't see them in the build dir. Same for `.pxd` files, it matters where they are located and whether there's an `__init__.py` in the same dir _and_ parent dirs. 

The alternative to avoid this behavior is to use `cython --module-name`.